### PR TITLE
Changed to v4.5 for Parallel LINQ. Added ReadAssemblies performance test.

### PR DESCRIPTION
**Pull request by tephe**
Opened this for discussion. What do you think of changing the target framework to 4.5 and using parallel linq on the enumerations? I think the changes could be minimal.

Made ILRepack.ReadAssemblies use AsParallel on my PC with 16 dlls from sharpdevelop and there's a decrease from 6 seconds to 4 seconds in execution. Will continue with RepackReferences.

### CODE REVIEW

File: ILRepack/ILRepack.cs [hunk 1]
```diff
@@ -74,15 +74,14 @@ public void Merge()
...
-        private void ReadInputAssemblies()
+        public void ReadInputAssemblies()
         {
             MergedAssemblyFiles = Options.InputAssemblies.SelectMany(ResolveFile).Distinct().ToList();
             OtherAssemblies = new List<AssemblyDefinition>();
-            // TODO: this could be parallelized to gain speed
+
             var primary = MergedAssemblyFiles.FirstOrDefault();
-            foreach (string assembly in MergedAssemblyFiles)
+            foreach (var result in MergedAssemblyFiles.Select(assembly => ReadInputAssembly(assembly, primary == assembly)).AsParallel())
```
Comment (hazzik) [inline ILRepack/ILRepack.cs]: Shouldn't it be

`.AsParallel().Select(assembly => ReadInputAssembly(assembly, primary == assembly))`

?

Comment (tephe) [inline ILRepack/ILRepack.cs]: Both ways are correct. Select() has deferred execution.

**Comment by gluck** [comment]
Nothing particular against the upgrade, but can't the TPL be used on 3.5 ?
http://www.nuget.org/packages/TaskParallelLibrary

File: ILRepack.Tests/ILRepack.Tests.csproj [hunk 2]
```diff
@@ -55,6 +58,15 @@
     <None Include="packages.config">
       <SubType>Designer</SubType>
     </None>
+    <None Include="sharpdevelop\ICSharpCode.SharpDevelop.BuildWorker120.exe.config" />
```
Comment (gluck) [inline ILRepack.Tests/ILRepack.Tests.csproj]: not expected here ? :)

Comment (gluck) [inline ILRepack.Tests/ILRepack.Tests.csproj]: Actually it is, the new test make use of them.
No other/better alternative to committing tons of assemblies/binairies in the repo ?

Comment (tephe) [inline ILRepack.Tests/ILRepack.Tests.csproj]: Yes, I just added this pull request to showcase and have a discussion about using parallel linq and performance test. Don't want to merge this. 

I also don't want to add all those binaries to the repository. The proper way would be to add a gradle script that gets these files from sharpdevelop (or any other repository with a lot of dlls? :)?. And I plan on doing that eventually. 

With the tests I want to spot bottlenecks and evaluate the performance boost of switching to parallel linq.

**Comment by tephe** [comment]
Didn't know about TaskParallelLibrary, thanks. Will add it instead of updating to v4.5.

File: ILRepack.Tests/ILRepack.Tests.csproj (hunk 2 above)
Comment (gluck) [inline ILRepack.Tests/ILRepack.Tests.csproj]: :+1: I was thinking about grabbing dlls from nuget, as it's *the* repository these days

**Comment by gluck** [comment]
I feel bad suggesting that, ILRepack can definitely be upgraded to 4.5.
(that being said, if TPL works fine in 3.5, we could start using it *today* you know where :smile: )

**Comment by tephe** [comment]
I think it will work okay, since it's developed by Microsoft, so I think it has to be the exact same code they used in 4.5. I'll switch to 4.5 then, then after we're done we can test the same code with 3.5 and TPL and keep it in a different branch if it's the same performance wise.

File: ILRepack.Tests/ILRepack.Tests.csproj (hunk 2 above)
Comment (tephe) [inline ILRepack.Tests/ILRepack.Tests.csproj]: You're right, it should be a lot easier to get them from nuget. But we need a project like sharpdevelop, which has quite a lot of dlls. I thought that nuget is more for standalone libraries. Will investigate more.

Comment (hazzik) [inline ILRepack.Tests/ILRepack.Tests.csproj]: I dont think we need any of these tests and test libraries. As they will only slow down the build, and not add any value.

**Pull request closed without merging**